examples/hyperpars_nb/script.jl

```diff
@@ -0,0 +1,344 @@
+    f = build_model(l, θ)
+    lik = f.lik
+    fz = f(l.x).fx
+    N = length(l.x)
+    qΩ = init_aux_posterior(lik, N)
+    Zygote.@ignore cavi!(fz, lik, l.x, l.y, l.m, l.S, qΩ; niter=1)
+    L = aug_elbo(lik, u_posterior(fz, l.m, l.S), l.x, l.y)
+    return -L
+end
+
+# In order to make this work with Zygote, we had to change from
+# `Centered` to `NonCentered` here. This deviates from the usual practice.
+
+function u_posterior(fz::AbstractGPs.FiniteGP, m::AbstractVector, S::AbstractMatrix)
+    return posterior(SparseVariationalApproximation(NonCentered(), fz, MvNormal(m, S)))
+end
+
+function aug_elbo(lik::GPLikelihoods.AbstractLikelihood, u_post, x::AbstractVector, y::AbstractVector)
+    qf = marginals(u_post(x))
+    qΩ = Zygote.@ignore aux_posterior(lik, y, qf)
+    term1 = expected_logtilt(lik, qΩ, y, qf)
+    term2 = ApproximateGPs.SparseVariationalApproximationModule._prior_kl(u_post.approx)
+    term3 = Zygote.@ignore begin
+        aux_kldivergence(lik, qΩ, y)
+    end
+    return term1 - term2 - term3
+end
+
+# This function had to be modified for the `NonCentered` `SparseVariationalApproximation`
+# This is an educated guess, hopefully it is correct!
+
+function cavi!(fz::AbstractGPs.FiniteGP, lik, x, y, m, S, qΩ; niter=10)
+    K = ApproximateGPs._chol_cov(fz)
+    κ = K.L
+    for _ in 1:niter
+        post_u = u_posterior(fz, m, S)
+        post_fs = marginals(post_u(x))
+        aux_posterior!(qΩ, lik, y, post_fs)
+        S .= inv(Symmetric(I + κ' * Diagonal(only(expected_auglik_precision(lik, qΩ, y))) * κ))
+        m .= S * κ' * only(expected_auglik_potential(lik, qΩ, y)) - κ \ mean(fz)
+    end
+    return m, S
+end
+
+function build_posterior(l::AVGPLoss, θ)
+    f = build_model(l, θ)
+    lik = f.lik
+    fz = f(l.x).fx
+    N = length(l.x)
+    qΩ = init_aux_posterior(lik, N)
+    cavi!(fz, lik, l.x, l.y, l.m, l.S, qΩ; niter=100)
+    post = u_posterior(fz, l.m, l.S)
+    posty = LatentGP(post, lik, 1e-6)
+    residuals = l.y - mean(posty(l.x))
+    return posty, residuals
+end
+
+
+# ## Overloads for posterior over Y
+# We can calculate the mean and variance of Y analytically
+function Statistics.mean(f::AbstractGPs.LatentFiniteGP{T1, NegativeBinomialLikelihood{NBParamFailure{T2}, LogisticLink}}) where {T1, T2}
+    r = f.lik.params.failures
+    m, c = mean(f.fx), var(f.fx)
+    return @. r * exp(m + c / 2)
+end
+
+function Statistics.var(f::AbstractGPs.LatentFiniteGP{T1, NegativeBinomialLikelihood{NBParamFailure{T2}, LogisticLink}}) where {T1, T2}
+    r = f.lik.params.failures
+    c = var(f.fx)
+    my = mean(f)
+    return @. my * (1 + my * ((1 + inv(r)) * exp(c) - 1))
+end
+
+# We approximate the posterior over Y by a truncated Gaussian distribution
+function AbstractGPs.marginals(
+    f::AbstractGPs.LatentFiniteGP{T1, NegativeBinomialLikelihood{NBParamFailure{T2}, LogisticLink}}
+) where {T1, T2}
+    m, c = mean(f), var(f)
+    return @. truncated(Normal(m, sqrt(c)), zero(eltype(m)), nothing)
+end
```

Comment:
Why would you need this? The marginals should not be truncated?

Reply:
This is a very rough approximation and makes sure that credible intervals are always nonnegative. This is now removed in favor of Monte Carlo estimates.